### ISSUE TITLE
On installation / sometimes restart or update of _other_ add-ons category is reset to "all"

### ISSUE THREAD

**Issue by asahukar** [RealRaven2000/QuickFolders#185]
I believe that once upon a time, Quick Folders would bring up the category displayed at the time of shutdown.

In one of the latest updates, that functionality has changed to displaying the 'All' category. 

Per our email conversation, you noted is that this gets disrupted and reset to "all" when you install a new version of QuickFolders, so far I haven't found a solution against this.

**Comment by RealRaven2000** [comment]
> Per our email conversation, you noted is that this gets disrupted and reset to "all" when you install a new version of QuickFolders, so far I haven't found a solution against this.

Unfortunately this doesn't happen always, so I might need your assistance with this. Let me just put down what I recorded in the email in case others want to participate:

the category is actually stored in for each (mail) tab via the internal Thunderbird session - it is done automatically when you select a category. within the function `QuickFolders.Interface.selectCategory( )`

If the tab is in the mode "folder" or "message" (3pane window or single message tab) I retrieve the tabInfo of the current tab and add a attribute "QuickFoldersCategory". You can double check this by activating the debug mode on QF advanced settings and then right-click the debug mode and enable
`extensions.quickfolders.debug.categories=true`

if I then select a category "Realraven", I can see in the Javascript error console (Ctrl+Shift+J to see it quickly) :

```
QuickFolders {CATEGORIES} 20:17:56.188 [3180 ms] selectCategory(RealRaven, false) quickfolders-util.js:1052:13
QuickFolders {CATEGORIES} 20:17:56.189 [1 ms] Selecting Categories: RealRaven… quickfolders-util.js:1052:13
QuickFolders {CATEGORIES} 20:17:56.190 [1 ms] set currentActiveCategories() Check menuitem: RealRaven quickfolders-util.js:1052:13
QuickFolders {CATEGORIES} 20:17:56.214 [7 ms] Successfully selected Category: RealRaven on mail tab [0] quickfolders-util.js:1052:13
```

then, if I simply restart Thunderbird I can see a line:

```
QuickFolders 20:16:2.894 [165 ms] restored tabs: RealRaven quickfolders-util.js:1057:13
QuickFolders 20:16:2.894 [0 ms] restore category RealRaven quickfolders-util.js:1057:13
QuickFolders 20:16:3.885 [78 ms] init: setting categories to RealRaven quickfolders-util.js:1057:13
QuickFolders {CATEGORIES} 20:16:3.885 [0 ms] selectCategory(RealRaven, undefined)
```

I might add some additional log info and drop another test version here later.

**Comment by RealRaven2000** [comment]
My current theory is that the session store function gets called after I restore the original function (which doesn't include the category) when QuickFolders is updated. This is the "wrapping" function that stores the information in the session after the original (Thunderbird built in) function is called:

```
  mailTabType.modes["folder"].persistTab = function(aTab) {
    let retval = orgPersist(aTab);
    if (retval) {
      util.logDebug("persist tab category: " + aTab.QuickFoldersCategory);
      retval.QuickFoldersCategory = aTab.QuickFoldersCategory; // add category from the tab to persisted info object
    }
    return retval; 
  }
```

So the question is, after QF gets shut down, may we store the tabs "QuickFoldersCategory" attribute? I guess if we do not restore the old Sessionstore object, then there is a change TB will still call this code, but it need to be safe (e.g. the util object may not exist anymore, and so it may fail).

**Comment by RealRaven2000** [comment]
So as a 1st workaround, I am not omitting the restauration of the old Session store object so that the Category will be still stored in the session in case Thunderbird restarts, provided QF has ran in the previous session at least once.

[QuickFolders-mx-5.7pre79.zip](https://github.com/RealRaven2000/QuickFolders/files/6874099/QuickFolders-mx-5.7pre79.zip)

Once you disable QF and then have  restarted Thunderbird it will of course not restore the category anymore, so the condition will not persist forever, but it should deal with the case of installing new versions or installation of other Add-ons.

**Comment by asahukar** [comment]
Axel,

I performed the following steps:

1. Disabled QF via the Thunderbird Extension Manager
2. Installed the contents of the .zip file above via the Extension Manager's "Install Add-on From File"
3. Shutdown Thunderbird
4. Started Thunderbird
5. Re-enabled QF
6. Shutdown Thunderbird
7. Started Thunderbird -- noted (Display All) was selected.
8. Selected 'AJ Gen'
9. Shutdown Thunderbird
10. Started Thunderbird -- again noted (Display All) was selected.

Is there any other information or settings data I can provide that might be useful to you?

Thanks again for your efforts here,

AJ

**Comment by RealRaven2000** [comment]
> I performed the following steps:
> 
> 1. Disabled QF via the Thunderbird Extension Manager
> 2. Installed the contents of the .zip file above via the Extension Manager's "Install Add-on From File"
> 3. Shutdown Thunderbird
> 4. Started Thunderbird
> 5. Re-enabled QF
> 6. Shutdown Thunderbird
> 7. Started Thunderbird -- noted (Display All) was selected.
> 8. Selected 'AJ Gen'
> 9. Shutdown Thunderbird
> 10. Started Thunderbird -- again noted (Display All) was selected.
> 
> Is there any other information or settings data I can provide that might be useful to you?
> 
> Thanks again for your efforts here,
> 
> AJ


If you could capture a debug log with debug mode and the setting
extensions.quickfolders.debug.categories=true
enabled, it might be helpful. Does it forget the categories every time you restart Thunderbird?

**Comment by asahukar** [comment]
I'd be happy to -- I assume instructions to enable debug mode are 
readily available?

AJ

**Comment by RealRaven2000** [comment]
> I'd be happy to -- I assume instructions to enable debug mode are readily available? AJ

yes here: 
https://quickfolders.org/support.html

Saving the error log has changed a little bit in the meantime you use "Export vbisible messages to clipboard" so I need to update that part.

**Comment by RealRaven2000** [comment]
Fixed in 5.7, [released on ATN](https://addons.thunderbird.net/thunderbird/addon/quickfolders-tabbed-folders/versions/5.7) today.

**Comment by asahukar** [comment]
Axel,

Hey there.

It took a while, but I can confirm that the issue does indeed to be 
fixed on my installation of Thunderbird.

My delay in replying was due to internet outage followed by power outage 
and has nothing to do with my lack of appreciation for your efforts.

Thanks again,

AJ
